src/events/tls.py

```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Event handler for related applications on the `certificates` relation interface."""
+import logging
+from typing import TYPE_CHECKING
+
+from charms.data_platform_libs.v0.upgrade import ActionEvent
+from charms.tls_certificates_interface.v1.tls_certificates import (
+    CertificateAvailableEvent,
+    TLSCertificatesRequiresV1,
+    generate_csr,
+    generate_private_key,
+)
+from ops.charm import RelationCreatedEvent, RelationJoinedEvent
+from ops.framework import EventBase, Object
+
+from core.utils import generate_password, parse_tls_file
+from literals import SUBSTRATE
+from managers.tls import TLSManager
+
+if TYPE_CHECKING:
+    from charm import ZooKeeperCharm
+
+logger = logging.getLogger(__name__)
+
+
+class TLSEvents(Object):
+    """Event handlers for related applications on the `certificates` relation interface."""
+
+    def __init__(self, charm):
```

Comment:
**suggestion** We are REALLY one-step away from avoiding circular dependency!!! We just need to provide as input of this class a `TLSManager` (and the `ClusterState` if we manage to get rid of the `State` in the `TLSManager` constructor - see another comment) and the `RollingOpsManager`, something like:

```
def __init__(self, charm: CharmBase, state: ClusterState, manager: TLSManager, rolling_ops: RollingOpsManager)
```

Substituting the references in the class would seem very easy and straightforward, e.g. `self.charm.state => self.state`

That would make me extremely happy, and most importantly, I would stop complaining about this :joy: 

That sounds something definitely worth addressing given those outcome, doesn't it? :stuck_out_tongue:

Reply:
We don't see the value gain as clear enough to warrant the complexity add, not to mention the divergence from all other Canonical charms. If it solved a real problem, we'll fix it, but for now this is a compromise that gives better delineation of responsibility.